tempesta_fw/apm.c

```diff
@@ -1151,14 +1496,318 @@ tfw_cfgop_apm_stats(TfwCfgSpec *cs, TfwCfgEntry *ce)
+
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_resp_code(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int i, code;
+	const char *val;
+
+	if (!ce->val_n) {
+		TFW_ERR_NL("No arguments found.\n");
+		return -EINVAL;
+	}
+	if (ce->attr_n) {
+		TFW_ERR_NL("Unexpected attributes\n");
+		return -EINVAL;
+	}
+	TFW_CFG_ENTRY_FOR_EACH_VAL(ce, i, val) {
+		if (tfw_cfgop_parse_http_status(val, &code)
+		    || tfw_cfg_check_range(code, HTTP_CODE_MIN, HTTP_CODE_MAX))
+		{
+			TFW_ERR_NL("Unable to parse http code value: '%s'\n",
+				   val);
+			return -EINVAL;
+		}
+		__set_bit(HTTP_CODE_BIT_NUM(code), tfw_hm_entry->codes);
+	}
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_resp_crc32(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	u32 crc32;
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+
+	if (tfw_cfg_parse_int(ce->vals[0], &crc32)) {
+		TFW_ERR_NL("Unable to parse crc32 value: '%s'\n", ce->vals[0]);
+		return -EINVAL;
+	}
+
+	tfw_hm_entry->crc32 = crc32;
+
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_timeout(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int timeout;
+
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+
+	if (tfw_cfg_parse_int(ce->vals[0], &timeout)) {
+		TFW_ERR_NL("Unable to parse http timeout value: '%s'\n",
+			   ce->vals[0]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_check_range(timeout, 1, USHRT_MAX))
+		return -EINVAL;
+
+	tfw_hm_entry->tmt = timeout;
+
+	return 0;
+}
+
+static void
+tfw_cfgop_cleanup_apm_hm(TfwCfgSpec *cs)
+{
+	TfwApmHM *hm, *tmp;
+
+	if (list_empty(&tfw_hm_list))
+		return;
+
+	list_for_each_entry_safe(hm, tmp, &tfw_hm_list, list) {
+		free_pages((unsigned long)hm->req, get_order(hm->reqsz));
+		kfree(hm->url);
+		list_del(&hm->list);
+		kfree(hm);
+	}
+	INIT_LIST_HEAD(&tfw_hm_list);
+}
+
+static TfwCfgSpec tfw_apm_hm_specs[] = {
+	{
+		.name		= "request",
+		.deflt		= "\"GET / HTTTP/1.0\r\n\r\n\"",
+		.handler	= tfw_cfgop_apm_hm_request,
+		.allow_none	= false,
+		.allow_repeat	= false,
+	},
+	{
+		.name		= "request_url",
+		.deflt		= NULL,
+		.handler	= tfw_cfgop_apm_hm_req_url,
+		.allow_none	= false,
+		.allow_repeat	= false,
+	},
+	{
+		.name		= "resp_code",
+		.deflt		= NULL,
+		.handler	= tfw_cfgop_apm_hm_resp_code,
+		.allow_none	= true,
+		.allow_repeat	= false,
+	},
+	{
+		.name		= "resp_crc32",
+		.deflt		= NULL,
+		.handler	= tfw_cfgop_apm_hm_resp_crc32,
+		.allow_none	= true,
+		.allow_repeat	= false,
+	},
+	{
+		.name		= "timeout",
+		.deflt		= NULL,
+		.handler	= tfw_cfgop_apm_hm_timeout,
+		.allow_none	= false,
+		.allow_repeat	= false,
+	},
+	{ 0 }
+};
+
 static TfwCfgSpec tfw_apm_specs[] = {
 	{
-		.name = "apm_stats",
-		.deflt = "window=300 scale=5",
-		.handler = tfw_cfgop_apm_stats,
-		.cleanup  = tfw_cfgop_cleanup_apm,
-		.allow_none = true,
-		.allow_repeat = false,
+		.name		= "apm_stats",
+		.deflt		= "window=300 scale=5",
+		.handler	= tfw_cfgop_apm_stats,
+		.cleanup	= tfw_cfgop_cleanup_apm,
+		.allow_none	= true,
+		.allow_repeat	= false,
+	},
+	{
+		.name		= "server_failover_http",
```

Comment:
The major thing about this patch is that this option should not be managed/processed by the APM module. I believe that architecturally is belongs to HTTP module. APM is about collecting, calculating, and managing various stats data, whereas the stats data itself my come from different sources and for different purposes. So this is a new and distinct stats characteristics, it's collected by the APM module, and the APM may tell the current state of various stats data, in this case - the health of a particular server.

In other words, these configuration options, I believe, belong to `socks_srv.c`, and half of the code should be in `http.c`, especially the parts that send health check http requests.